databricks/koalas/base.py

```diff
@@ -94,27 +200,43 @@ def column_op(f):
 
     @wraps(f)
     def wrapper(self, *args):
+        from databricks.koalas.indexes import Index
+        from databricks.koalas.series import Series
+
         # It is possible for the function `f` takes other arguments than Spark Column.
         # To cover this case, explicitly check if the argument is Koalas Series and
         # extract Spark Column. For other arguments, they are used as are.
         cols = [arg for arg in args if isinstance(arg, IndexOpsMixin)]
 
-        # TODO: support ops between different types.
-        assert all(
-            type(self) == type(col) for col in cols
-        ), "All arguments must be the same types: [{}, {}]".format(
-            type(self).__name__, ", ".join(type(col).__name__ for col in cols)
-        )
-
-        if all(not self._need_alignment_for_column_op(col) for col in cols):
+        if all(not need_alignment_for_column_op(self, col) for col in cols):
             # Same DataFrame anchors
             args = [arg.spark.column if isinstance(arg, IndexOpsMixin) else arg for arg in args]
             scol = f(self.spark.column, *args)
             scol = booleanize_null(self.spark.column, scol, f)
 
-            index_ops = self._with_new_scol(scol)
+            if isinstance(self, Series) or not any(isinstance(col, Series) for col in cols):
+                index_ops = self._with_new_scol(scol)
+            else:
+                kser = next(col for col in cols if isinstance(col, Series))
+                index_ops = kser._with_new_scol(scol)
+        elif get_option("compute.ops_on_diff_frames"):
+            if isinstance(self, Series) and all(isinstance(col, Series) for col in cols):
+                index_ops = align_diff_series(f, self, *args)
+            else:
+                # This could cause as many counts, reset_index calls, joins for combining
+                # as the number of `Index`s in `args`. So far it's fine since we can assume the ops
+                # only work between at most two `Index`s. We might need to fix it in the future.
+
+                self_len = len(self)
+                if any(len(col) != self_len for col in args if isinstance(col, IndexOpsMixin)):
+                    raise ValueError("operands could not be broadcast together with shapes")
+
+                if isinstance(self, Index) and all(isinstance(col, Index) for col in cols):
+                    index_ops = align_diff_index(f, self, *args)
+                else:
+                    index_ops = align_diff_index_ops(f, self, *args)
         else:
-            index_ops = self._align_and_column_op(f, *args)
+            raise ValueError(ERROR_MESSAGE_CANNOT_COMBINE)
 
         if not all(self.name == col.name for col in cols):
             index_ops = index_ops.rename(None)
```

Comment:
I am wondering whether `index_ops`'s name should always be `None`.

Reply:
And I may have some concern about what `index_ops` means. It seems to be the result we return finally.

Reply:
@xinrong-databricks When the name is same, seems they keep it in pandas.

```python
>>> pd.Series([1, 2, 3], name="Koalas") + pd.Index([2, 3, 4], name="Koalas")
0    3
1    5
2    7
Name: Koalas, dtype: int64
```

Reply:
Got it, thanks!